python/mlx/optimizers/schedulers.py

```diff
@@ -84,3 +85,87 @@ def scheduler(step):
         return init * decay
 
     return scheduler
+
+
+class ScheduleJoiner:
+    r"""Make a schedule joiner
+
+    Instanciated with a list of schedules and an increasing list of 0-based steps, each of which indicates
+    when the next schedule will be used.  The instanciated object is a callable schedule that is a
+    concatenation of the schedules at the given step boundaries
+
+    So, the first item in the schedule list will be used until the first boundary, then the next schedule
+    will be used until the next boundary, and so on.
+
+    Args:
+        schedules (list of schedules): The schedules to join
+        boundaries (list of integers): The list of steps that mark the use of the next schedule (in order)
+
+    Example:
+        >>> warmup_schedule = optim.linear_warmup(100, finish=1e-1)
+        >>> cosine_schedule = optim.cosine_decay(1e-1, 200)
+        >>> lr_schedule = ScheduleJoiner([warmup_schedule, cosine_schedule], [101])
+        >>> optimizer = optim.Adam(learning_rate=lr_schedule)
+        >>> optimizer.learning_rate
+        array(0.0, dtype=float32)
+        >>> for _ in range(101): optimizer.update({}, {})
+        ...
+        >>> optimizer.learning_rate
+        array(1e-5, dtype=float32)
+
+    """
+
+    def __init__(self, schedules: List[Callable], boundaries: List[int]):
+        self.schedules = schedules
+        self.boundaries = boundaries
+
+    def __call__(self, step: int):
+        if step < self.boundaries[0] or not self.boundaries:
```

Comment:
We need to rethink this approach. We shouldn't be doing control flow based on the value of `step` which is an `mx.array` because it won't work well with function transformations..